src/api/app/models/staging_project.rb

```diff
@@ -0,0 +1,142 @@
+# The staging project is the place where we collect a selection of packages and test, build and review them.
+class StagingProject < Project
+  has_many :staged_requests, class_name: 'BsRequest', foreign_key: :staging_project_id, dependent: :nullify
+  has_many :status_reports, through: :repositories
+  belongs_to :staging_workflow, inverse_of: :staging_projects
+
+  def staging_identifier
+    name[/.*:Staging:(.*)/, 1]
+  end
+
+  def untracked_requests
+    requests_to_review - staged_requests
+  end
+
+  # The difference between staged requests and requests to review is that staged requests are assigned to the staging project.
+  # The requests to review are requests which are not related to the staging project (unless they are also staged).
+  # They simply need a review from the maintainers of the staging project.
+  def requests_to_review
+    @requests_to_review ||= BsRequest.with_open_reviews_for(by_project: name)
+  end
+
+  def building_repositories
+    set_buildinfo unless @building_repositories
+    @building_repositories
+  end
+
+  def broken_packages
+    set_buildinfo unless @broken_packages
+    @broken_packages
+  end
+
+  def missing_reviews
+    return @missing_reviews if @missing_reviews
+
+    @missing_reviews = []
+    attribs = [:by_group, :by_user, :by_project, :by_package]
+
+    (requests_to_review + staged_requests).uniq.each do |request|
+      request.reviews.each do |review|
+        next if review.state.to_s == 'accepted' || review.by_project == name
+        # FIXME: this loop (and the inner if) would not be needed
+        # if every review only has one valid by_xxx.
+        # I'm keeping it to mimic the python implementation.
+        # Instead, we could have something like
+        # who = review.by_group || review.by_user || review.by_project || review.by_package
```

Comment:
I guess this is c&p right? What is the problem with the or? 

cc @coolo

Reply:
@ChrisBr Yes this is a copy and paste.

Reply:
Since a review's `by_user`, `by_group`, `by_project` and `by_package` can all be defined (but they don't have to), we need to go through all of them. With the `or`, we would only check the first one which is defined (even though there is the possibility of having another `by_...` defined afterwards that`.

Reply:
from review.rb:
```
    if validate_reviewer_fields
      errors.add(:base, 'it is not allowed to have more than one reviewer entity: by_user, by_group, by_project, by_package')
    end

Reply:
if by_package is set, by_project is the project of it - there is basically no way to have by_package *without* by_project.

But by_user and by_group are mutually exclusive and 
```
MariaDB [obs_production]> select distinct number,by_package,by_project from reviews v join bs_requests r on r.id=v.bs_request_id where by_user is not null and by_group is not null;
+--------+------------+------------+
| number | by_package | by_project |
+--------+------------+------------+
| 491574 | NULL       | NULL       |
+--------+------------+------------+
```

And https://build.opensuse.org/request/show/491574 looks like an exercise In free love (if you don't know the reference, you're lost on music :)